br/pkg/lightning/restore/check_info.go

```diff
@@ -399,535 +180,64 @@ func (rc *Controller) estimateSourceData(ctx context.Context) (int64, int64, boo
-	dbInfo, ok := dbInfos[tableInfo.DB]
-	if ok {
-		t, ok := dbInfo.Tables[tableInfo.Name]
-		if ok {
-			if tableCheckPoint.TableID > 0 && tableCheckPoint.TableID != t.ID {
-				msgs = append(msgs, fmt.Sprintf("TiDB Lightning has detected tables with illegal checkpoints. To prevent data loss, this run will stop now,"+
-					"please run command \"./tidb-lightning-ctl --checkpoint-remove='%s' --config=...\""+
-					"You may also run `./tidb-lightning-ctl --checkpoint-error-destroy=all --config=...` to start from scratch,"+
-					"For details of this failure, read the log file from the PREVIOUS run",
-					uniqueName))
-				return msgs, false
-			}
-		}
-	}
-
-	var permFromCheckpoint []int
-	var columns []string
-	for _, eng := range tableCheckPoint.Engines {
-		if len(eng.Chunks) > 0 {
-			chunk := eng.Chunks[0]
-			permFromCheckpoint = chunk.ColumnPermutation
-			columns = chunk.Chunk.Columns
-			if filepath.Dir(chunk.FileMeta.Path) != rc.cfg.Mydumper.SourceDir {
-				message := fmt.Sprintf("chunk checkpoints path is not equal to config"+
-					"checkpoint is %s, config source dir is %s", chunk.FileMeta.Path, rc.cfg.Mydumper.SourceDir)
-				msgs = append(msgs, message)
-			}
-		}
-	}
-	if len(columns) == 0 {
-		log.FromContext(ctx).Debug("no valid checkpoint detected", zap.String("table", uniqueName))
-		return nil, false
-	}
-	info := dbInfos[tableInfo.DB].Tables[tableInfo.Name]
-	if info != nil {
-		permFromTiDB, err := parseColumnPermutations(info.Core, columns, nil, log.FromContext(ctx))
+		result, err := theChecker.Check(ctx)
 		if err != nil {
-			msgs = append(msgs, fmt.Sprintf("failed to calculate columns %s, table %s's info has changed,"+
-				"consider remove this checkpoint, and start import again.", err.Error(), uniqueName))
+			return errors.Trace(err)
 		}
-		if !reflect.DeepEqual(permFromCheckpoint, permFromTiDB) {
-			msgs = append(msgs, fmt.Sprintf("compare columns perm failed. table %s's info has changed,"+
-				"consider remove this checkpoint, and start import again.", uniqueName))
+		if result != nil {
+			rc.checkTemplate.Collect(result.Severity, result.Passed, result.Message)
 		}
 	}
-	return msgs, false
-}
 
-// hasDefault represents col has default value.
-func hasDefault(col *model.ColumnInfo) bool {
-	return col.DefaultIsExpr || col.DefaultValue != nil || !mysql.HasNotNullFlag(col.GetFlag()) ||
-		col.IsGenerated() || mysql.HasAutoIncrementFlag(col.GetFlag())
-}
-
-func (rc *Controller) readFirstRow(ctx context.Context, dataFileMeta mydump.SourceFileMeta) (cols []string, row []types.Datum, err error) {
-	cols, rows, err := rc.preInfoGetter.ReadFirstNRowsByFileMeta(ctx, dataFileMeta, 1)
-	if err != nil {
-		return nil, nil, err
-	}
-	if len(rows) > 0 {
-		return cols, rows[0], nil
-	} else {
-		return cols, []types.Datum{}, nil
-	}
-}
-
-// SchemaIsValid checks the import file and cluster schema is match.
-func (rc *Controller) SchemaIsValid(ctx context.Context, tableInfo *mydump.MDTableMeta) ([]string, error) {
```

Comment:
Extracted as `(*schemaCheckItem).Check()`, in `br/pkg/lightning/restore/precheck_impl.go`, L681